gateway-ha/src/main/java/io/trino/gateway/ha/router/BaseRoutingManager.java

```diff
@@ -0,0 +1,306 @@
+    public void setRoutingGroupForQueryId(String queryId, String routingGroup)
+    {
+        queryIdRoutingGroupCache.put(queryId, routingGroup);
+    }
+
+    /**
+     * Performs routing to a default backend.
+     */
+    public ProxyBackendConfiguration provideDefaultBackendConfiguration(String user)
+    {
+        List<ProxyBackendConfiguration> backends = gatewayBackendManager.getActiveDefaultBackends();
+        backends.removeIf(backend -> isBackendNotHealthy(backend.getName()));
+        return selectBackend(backends, user).orElseThrow(() -> new IllegalStateException("Number of active backends found zero"));
+    }
+
+    /**
+     * Performs routing to a given cluster group. This falls back to a default backend, if no scheduled
+     * backend is found.
+     */
+    @Override
+    public ProxyBackendConfiguration provideBackendConfiguration(String routingGroup, String user)
+    {
+        List<ProxyBackendConfiguration> backends = gatewayBackendManager.getActiveBackends(routingGroup);
+        backends.removeIf(backend -> isBackendNotHealthy(backend.getName()));
+        return selectBackend(backends, user).orElseGet(() -> provideDefaultBackendConfiguration(user));
+    }
+
+    protected abstract Optional<ProxyBackendConfiguration> selectBackend(List<ProxyBackendConfiguration> backends, String user);
+
+    /**
+     * Performs cache look up, if a backend not found, it checks with all backends and tries to find
+     * out which backend has info about given query id.
+     */
+    @Nullable
+    @Override
+    public String findBackendForQueryId(String queryId)
+    {
+        String backendAddress = null;
+        try {
+            backendAddress = queryIdBackendCache.get(queryId);
+        }
+        catch (ExecutionException e) {
+            log.warn("Exception while loading queryId from cache %s", e.getLocalizedMessage());
+        }
+        return backendAddress;
+    }
+
+    @Nullable
+    @Override
+    public String findExternalUrlForQueryId(String queryId)
+    {
+        String externalUrl = null;
+        try {
+            externalUrl = queryIdExternalUrlCache.get(queryId);
+        }
+        catch (ExecutionException e) {
+            log.warn("Exception while loading queryId from cache %s", e.getLocalizedMessage());
+        }
+        return externalUrl;
+    }
+
+    /**
+     * Looks up the routing group associated with the queryId in the cache.
+     * If it's not in the cache, look up in query history
+     */
+    @Nullable
+    @Override
+    public String findRoutingGroupForQueryId(String queryId)
+    {
+        String routingGroup = null;
+        try {
+            routingGroup = queryIdRoutingGroupCache.get(queryId);
+        }
+        catch (ExecutionException e) {
+            log.warn("Exception while loading queryId from routing group cache %s", e.getLocalizedMessage());
+        }
+        return routingGroup;
+    }
+
+    protected void updateBackEndHealth(List<ClusterStats> stats)
```

Comment:
can we delete this method ? 

It get confusing as we have 3 methods that update the state in cache.  
You move this method logic into the interface method `public void updateClusterStats` . This would simplify and we avoid method overloading of interface method.

Reply:
Done